### PR TITLE
Try overhaul and implicits

### PR DESCRIPTION
**Pull request by siroky**
For Option, it works really nicely.

For Try, it is not perfect, because of two facts:
1) We have `Try<A>` and `Try<A, E>` and there is no way of turning `Error<InvalidOperationException>` into Try<A> because of type invariance of Error. So I decided to use `Error<E>` for `Try<A, E>` and `Exception` for `Try<A>`. It's not as nice but works pretty well.
2) Second problem which I consider a bit more serious is, that as soon as you start using Match and nested expressions, you have to explicitly put in types, so that the compiler knows the resulting type. This can be seen inside `Try<A>` for example.

I still don't like the distinction between two versions of Try and consider actually getting rid of `Try<A>` and instead implement extension methods for `Try<A, E> where E : Exception`. On the one hand, typing would be more elaborate, on the other hand, we'd get more specific type signatures and thanks to implicit conversions, usage would be even better than now.

### CODE REVIEW

**Comment by starychfojtu** [comment]
I would definitely vote for removing Try<A>, speaking of our codebase, we already have misuses for Try<A> like http client, which always returns Try<Response, HttpClientException>, but its signature is still Try<A>. And moving forward, we should aim for limiting the usage of weakly typed errors, So I am all for this.

File: src/FuncSharp/Extensions/BooleanExtensions.cs [hunk 1]
```diff
@@ -23,7 +23,10 @@ public static void Match(this bool b, Action<Unit> ifTrue, Action<Unit> ifFalse)
...
         public static Try<T> ToTry<T>(this bool b, Func<Unit, T> ifTrue, Func<Unit, Exception> ifFalse)
         {
-            return b ? Try.Success<T>(ifTrue(Unit.Value)) : Try.Error<T>(ifFalse(Unit.Value));
+            return b.Match<Try<T>>(
+                t => Try.Success(ifTrue(Unit.Value)),
+                f => Try.Exception(ifFalse(Unit.Value))
+            );
         }
 
         public static Try<T, E> ToTry<T, E>(this bool b, Func<Unit, T> ifTrue, Func<Unit, E> ifFalse)
```
Comment (starychfojtu) [inline src/FuncSharp/Extensions/BooleanExtensions.cs]: This cannot be changed to use Try.Success and Try.Error in its implementation ?

File: src/FuncSharp/Try/Try.cs [hunk 2]
```diff
@@ -49,83 +47,67 @@ public static Try<A> Create<A>(Func<Unit, A> f)
...
-            return new Try<A, E>(error);
+            return new TryError<E>(error);
         }
 
         /// <summary>
-        /// Creates a new try with an exception result.
+        /// Creates a new try with an error result.
         /// </summary>
-        public static Try<A> Error<A>(Exception exception)
+        public static Try<A, E> Error<A, E>(E error)
```
Comment (starychfojtu) [inline src/FuncSharp/Try/Try.cs:74]: I think there might be problem with inference among these two methods, have you encountered any while writing it ?

File: src/FuncSharp/Try/Try.cs [hunk 3]
```diff
@@ -649,38 +585,21 @@ public A Get()
...
-                {
-                    var exception = e.SingleOption();
-                    if (exception.NonEmpty)
-                    {
-                        ExceptionDispatchInfo.Capture(exception.Get()).Throw();
-                    }
-
-                    throw new AggregateException(e);
-                }
+                e => throw e
```
Comment (starychfojtu) [inline src/FuncSharp/Try/Try.cs]: This will forget the stacktrace, previously the stack trace worked for single exception

**Review by starychfojtu — APPROVED** [review]

File: src/FuncSharp/Try/Try.cs (hunk 2 above)
Comment (siroky) [reply src/FuncSharp/Try/Try.cs:74]: This version is for cases, where you explicitly want to specify full type without the implicit conversions. Might be useful in some more complicated places where the compiler cannot infer it easily.

**Pull request merged**